src/comgr.cpp

```diff
@@ -0,0 +1,503 @@
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_EXECUTABLE);
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_BYTES);
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_FATBIN);
+    }
+    return "<Unknown data kind>";
+}
+
+static std::string to_string(const amd_comgr_action_kind_t val)
+{
+    switch(val)
+    {
+        CASE_TO_STRING(AMD_COMGR_ACTION_SOURCE_TO_PREPROCESSOR);
+        CASE_TO_STRING(AMD_COMGR_ACTION_ADD_PRECOMPILED_HEADERS);
+        CASE_TO_STRING(AMD_COMGR_ACTION_COMPILE_SOURCE_TO_BC);
+        CASE_TO_STRING(AMD_COMGR_ACTION_ADD_DEVICE_LIBRARIES);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_BC_TO_BC);
+        CASE_TO_STRING(AMD_COMGR_ACTION_OPTIMIZE_BC_TO_BC);
+        CASE_TO_STRING(AMD_COMGR_ACTION_CODEGEN_BC_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_CODEGEN_BC_TO_ASSEMBLY);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_RELOCATABLE_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_RELOCATABLE_TO_EXECUTABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_ASSEMBLE_SOURCE_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_RELOCATABLE_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_EXECUTABLE_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_BYTES_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_COMPILE_SOURCE_TO_FATBIN);
+    }
+    return "<Unknown action kind>";
+}
+
+static bool PrintVersion()
+{
+    std::size_t major = 0;
+    std::size_t minor = 0;
+    (void)amd_comgr_get_version(&major, &minor);
+    MIOPEN_LOG_I("comgr v." << major << '.' << minor);
+    return true;
+}
+
+static std::string GetStatusText(const amd_comgr_status_t status)
+{
+    const char* reason = nullptr;
+    if(AMD_COMGR_STATUS_SUCCESS != amd_comgr_status_string(status, &reason))
+        reason = "<Unknown>";
+    return std::string(reason) + " (" + std::to_string(static_cast<int>(status)) + ')';
+}
+
+static void LogOptions(const char* options[], size_t count)
+{
+#if DEBUG_DETAILED_LOG
+    if(miopen::IsLogging(miopen::LoggingLevel::Info))
+    {
+        std::ostringstream oss;
+        for(std::size_t i = 0; i < count; ++i)
+            oss << options[i] << '\t';
+        MIOPEN_LOG_I(oss.str());
+    }
+#else
+    std::ignore = options;
+    std::ignore = count;
+#endif
+}
+
+class Dataset;
+static std::string GetLog(const Dataset& dataset, bool catch_comgr_exceptions = false);
+
+struct ComgrError : std::exception
+{
+    amd_comgr_status_t status;
+    std::string text;
+
+    ComgrError(const amd_comgr_status_t s) : status(s) {}
+    ComgrError(const amd_comgr_status_t s, const std::string& t) : status(s), text(t) {}
+    const char* what() const noexcept override { return text.c_str(); }
+};
+
+struct ComgrOwner
+{
+    ComgrOwner(const ComgrOwner&) = delete;
+
+    protected:
+    ComgrOwner() {}
+    ComgrOwner(ComgrOwner&&) = default;
+    [[noreturn]] void Throw(const amd_comgr_status_t s) const { throw ComgrError{s}; }
```

Comment:
Why arent we using `MIOPEN_THROW`? Using `MIOPEN_THROW` can report back the source line numbers where the exception was thrown. Usually we write a function to convert the error enum into a message and then pass that `MIOPEN_THROW`.

Reply:
There exceptions are not intended to report "normal" miopen errors. These report kernel build errors (like the ones that offline compiler prints onto console). I need to catch all these errors within comgr module.

This comgr stuff is a "library within a library" that provides our HIP backend with functionality similar to what `clBuildProgram()` gives us for free when we use OpenCL backend. If we use MIPEN_THROW here then this would be like if clBuildProgram() throws miopen-specific exceptions from within its internals.

Reply:
But its not a separate library so we should use `MIOPEN_THROW`. If or when we move it to a separate library we can change it to another throw mechanism.

Reply:
I do not get this. Why the library is limited to use only one type of exceptions? Reasons?

Reply:
Because at the API level it catches these exceptions and turn them into proper error codes. In addition it will show where in the source code the exception is thrown so we can have a better idea of where the error is happening.

If you want a different exception type, you could extend miopen's exception, but we need to add a macro for throwing different exceptions:

```cpp
#define MIOPEN_THROW_EX(ex, ...) throw ex(__VA_ARGS__).SetContext(__FILE__, __LINE__)
```

Reply:
This is feasible but seems useless to me. Why:

> In addition it will show where in the source code the exception is thrown 

File/line information is not needed here (and even considered harmful). 

> Because at the API level it catches these exceptions and turn them into proper error codes. 

These exceptions are not allowed to escape comgr module. We shall consider it as an internals of the compiler driver. The main purpose of this exception is to prevent resource leakage (comgr handles). The side functionality is to hold status codes and error messages received from comgr.

Reply:
Explanatory comment added to the source code.